Initial implementation

dhontecillas opened the pull request:
Please, take a look at the [README.md](https://github.com/krakend/krakend-otel/blob/60bd80de002ac9d10897b39941296d4560deead9/README.md) for information about how to configure the library, and at [implementation details](./doc/implementation_details.md) to have an initial explanation of the packages that compose the library.

The provided example is currently working, and easy to check the reported traces, however some work must be done, to create some dashboard, and check the metrics.

Also, there are still pending tests to be written. 

But this implementation is ready to start receiving some feedback.

dhontecillas committed:
initial implementation of otel metrics

taik0 commented on README.md:
```diff
@@ -1,72 +1,245 @@
+
+In order to configure the open telemetry stack to instrument the API Gateway, a 
+new entry must be added to the `ExtraConfig` root propertry of `ServiceConfig` 
+using the `telemetry/opentelemery` key, with the `krakend-otel`'s configuration.
```
falta una **t** en  opentelemery

taik0 commented on README.md:
```diff
@@ -1,72 +1,245 @@
+
+In the configuration we find the following root entries:
+
+- `service_name`: to provide a custom name for this service. However if there is 
```
Yo optaría por hacerlo al revés. Si `service_name` está definido se usa, si no existe se usa el `ServiceConfig.Name` y si tampoco existe, un default tipo `KrakenD`.

taik0 commented on README.md:
```diff
@@ -1,72 +1,245 @@
+
+At the router level we have 3 main options:
+
+- `metrics`: boolean to enable / disable if we want to report metrics for this layer
```
En opencensus usabamos lo contrario. `disable_metrics` o `disable_traces` para que el default de la struct sea que esté activado (que entiendo es lo que nos interesa).

taik0 commented on config/lura.go:
```diff
@@ -0,0 +1,70 @@
+		return nil, err
+	}
+
+	if cfg.Layers == nil {
```
If we customize a layer, the rest will be disabled, that can be confusing since not setting anything enables everything by default. May be, each layer should be checked independently. We can also reverse the option so the default struct value (`false`) mimics this behaviour.

taik0 commented on exporter/exporter.go:
```diff
@@ -0,0 +1,166 @@
+		}
+		if ss, ok := i.(SpanExporter); ok && ss != nil {
+			s[name] = ss
+		} else if mm, ok := i.(MetricReader); ok {
```
`mm` can't be nil?

dhontecillas replied:
should not happen, but since we are not sure, I will add the check as in line 91 , good catch!

kpacha commented:
I'd rather move the `skip_path` option to the router instead of having it at the backend, so a set of pipes could be ignored

alombarte commented on README.md:
````diff
@@ -1,72 +1,304 @@
+
+```json
+"exporters": {
+    "local_prometheus": {
````
Thinking in external tools and config validation, I would rather avoid having a "free text" key, and convert exporters to an array instead. Something like:
```
"exporters": [
{
  "name": "my_prometheus_fashion_Exporter",
  "kind": "prometheus",
  "config": {
     "port": 9090,
     "process_metrics": true,
     "go_metrics": true
   }
 },
....
]```

dhontecillas commented:
A better format for the exporters would be:

```json
exporters": {
                    "prometheus": [
                        {
                            "name": "remote_prometheus",
                            "port": 9092,
                            "process_metrics": true,
                            "go_metrics": true
                        },
                        {
                            "name": "local_prometheus",
                            "port": 9092,
                            "process_metrics": true,
                            "go_metrics": true
                        }]
                }

```

Harnoor7 commented on exporter/exporter.go:
```diff
@@ -0,0 +1,175 @@
+// MetricReader is the interface required in order to
+// export metrics.
+type MetricReader interface {
+	MetricReader() sdkmetric.Reader
```
Why we are not using `sdkmetric.Exporter` here?

noob-se7en commented on exporter/exporter.go:
```diff
@@ -0,0 +1,175 @@
+	s := make(map[string]SpanExporter)
+	var errList []error
+
+	ctx := context.Background()
```
nit, we should take context from params

noob-se7en commented on doc/TODO.md:
```diff
@@ -0,0 +1,7 @@
+# TODO
```
shutting down meter provider gracefully is also TODO

noob-se7en commented:
@dhontecillas any reason for not using [otel-go net/http](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/net/http/otelhttp) standard instrumentation support library?

dhontecillas commented:
> @dhontecillas any reason for not using [otel-go net/http](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/net/http/otelhttp) standard instrumentation support library?

@Harnoor7 Main reason is that I was not aware of it. But now that I've taken a first look at it, I think the library is missing some stuff (like metrics for the http client data, or keeping track of the readed bytes from the client). I might be wrong, and I will carefully review the differences there. The focus now is on having this PR in a usable state, then I will review and check we can have the same metrics / traces with the contrib library .. in that case I will use what I can from there, and if I miss something, we might think about trying to make a PR there.

I will keep an eye on it. Thanks for pointing out.

noob-se7en commented:
@dhontecillas  Got it. Yes, you are correct client-side metric instrumentation is missing, which will be added probably in the next release. There is an active PR for it: [https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4707](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4707)

dhontecillas committed:
log the otel errors in lura format

dhontecillas committed:
remove leftover of old code

dhontecillas committed:
finalize trace when connection is hijacked / upgraded to grpc, websocket...

deepsource-autofix[bot] committed:
style: format code with Go fmt and Gofumpt

This commit fixes the style issues introduced in 33d60ff according to the output
from Go fmt and Gofumpt.

Details: https://github.com/krakend/krakend-otel/pull/1

dhontecillas committed:
fix pipe naming in comment

dhontecillas committed:
Merge branch 'initial_implementation' of github.com:krakend/krakend-otel into initial_implementation

dhontecillas committed:
rename attribute krakend.endpoint krakend.endpoint.route

dhontecillas committed:
fix how backend hosts attributes is reported

dhontecillas committed:
fix attributes at the proxy level

dhontecillas committed:
update to stable version of otel metrics otel library

dhontecillas committed:
expose providers in the state

dhontecillas committed:
expose access to the providers with the interface
